substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/graal/snippets/NonSnippetLowerings.java

```diff
@@ -265,6 +265,13 @@ public void lower(FixedNode node, LoweringTool tool) {
                     if (!SubstrateBackend.shouldEmitOnlyIndirectCalls()) {
                         loweredCallTarget = graph.add(new DirectCallTargetNode(parameters.toArray(new ValueNode[parameters.size()]),
                                         callTarget.returnStamp(), signature, targetMethod, callType, invokeKind));
+                    } else if (!targetMethod.hasCodeOffsetInImage()) {
```

Comment:
Don't you need to have this `if` before the check of `shouldEmitOnlyIndirectCalls`? We might want to go back to direct calls in JIT compiled code at some point. This means that you need to implement `HostedMethod.hasCodeOffsetInImage` as `return true`.

Reply:
Not sure I fully understand. Isn't `hasCodeOffsetInImage` (and `getCodeOffsetInImage`) only relevant for indirect calls? Even if we emit direct calls in JIT compiled code, we would not use `getCodeOffsetInImage`, would we? And even if we would, at which point in the compilation pipeline would we call it? Because the new check is really only about avoiding the eager call to `getCodeOffsetInImage` for calls we expect to fold away.

Or asked differently, if I would implement `HostedMethod.hasCodeOffsetInImage` as `return true`, how would `HostedMethod.getCodeOffsetInImage` look like?

Reply:
Before the `shouldEmitOnlyIndirectCalls` check was introduced, direct calls were emitted in JIT methods and `RuntimeCodeInstaller` did the patching, i.e., called `getCodeOffsetInImage` for the call target. Now you can argue of course that this happened at the very end of compilation, so your unreachable invoke would have been removed by then. This means compilation would actually work.

So from that point of view, you can argue again that the current place of your check is appropriate...

> if I would implement HostedMethod.hasCodeOffsetInImage as return true, how would HostedMethod.getCodeOffsetInImage look like?

`getCodeOffsetInImage` would still be `unimplemented()`. Because it would still be unused even when `hasCodeOffsetInImage` returns true.

So considering all of this, it really does not matter where the `if` is placed. So I'm fine with keeping it as it is.